python/kwiver/sprokit/tests/sprokit/adapters/test-adapter_data_set.py

```diff
@@ -0,0 +1,258 @@
+    add_get_helper(ads, ads.add_track_set, ads.get_port_data_track_set, kvt.TrackSet(), "track_set")
+    add_get_helper(ads, ads.add_object_track_set, ads.get_port_data_object_track_set, kvt.ObjectTrackSet(), "object_track_set")
+
+# Next some bound native C++ types
+def test_add_get_cpp_types():
+    from kwiver.sprokit.adapters import adapter_data_set
+    from kwiver.sprokit.pipeline import datum
+
+    ads = adapter_data_set.AdapterDataSet.create()
+    add_get_helper(ads, ads.add_double_vector, ads.get_port_data_double_vector, datum.VectorDouble([3.14, 4.14]), "double_vector")
+    add_get_helper(ads, ads.add_string_vector, ads.get_port_data_string_vector, datum.VectorString(["s00", "s01"]), "string_vector")
+    add_get_helper(ads, ads.add_uchar_vector, ads.get_port_data_uchar_vector, datum.VectorUChar([100, 101]), "uchar_vector")
+
+# Now try creating datums of these bound types
+def test_add_get_cpp_types_with_datum():
+    from kwiver.sprokit.adapters import adapter_data_set
+    from kwiver.sprokit.pipeline import datum
+
+    ads = adapter_data_set.AdapterDataSet.create()
+    add_get_helper(ads, ads.add_datum, ads.get_port_data_double_vector, datum.new_double_vector(datum.VectorDouble([6.3, 8.9])), "datum_double_vector")
+    add_get_helper(ads, ads.add_datum, ads.get_port_data_string_vector, datum.new_string_vector(datum.VectorString(["foo", "bar"])), "datum_string_vector")
+    add_get_helper(ads, ads.add_datum, ads.get_port_data_uchar_vector, datum.new_uchar_vector(datum.VectorUChar([102, 103])), "datum_uchar_vector")
+
+# Next kwiver vital types
+def test_add_get_vital_types():
+    from kwiver.vital import types as kvt
+    from kwiver.sprokit.adapters import adapter_data_set
+    from kwiver.vital import types as kvt
+
+    ads = adapter_data_set.AdapterDataSet.create()
+    add_get_helper(ads, ads.add_bounding_box, ads.get_port_data_bounding_box, kvt.BoundingBox(1, 1, 2, 2), "bounding_box")
+    add_get_helper(ads, ads.add_timestamp, ads.get_port_data_timestamp, kvt.Timestamp(), "timestamp")
+    add_get_helper(ads, ads.add_f2f_homography, ads.get_port_data_f2f_homography, kvt.F2FHomography(1), "f2f_homography")
+
+# Now test overwriting
+def test_overwrite():
+    from kwiver.vital import types as kvt
+    from kwiver.sprokit.adapters import adapter_data_set
+    from kwiver.sprokit.pipeline import datum
+
+    OVERWRITE_PORT = "test_overwrite_port"
+    ads = adapter_data_set.AdapterDataSet.create()
+
+    # Overwriting with same datum
+    ads.add_datum(OVERWRITE_PORT, datum.new("d2"))
+    overwrite_helper(ads.add_datum, ads.get_port_data, datum.new("d3"), "datum_string", OVERWRITE_PORT)
+
+    # Overwriting with completely different types
+    overwrite_helper(ads.add_datum, ads.get_port_data, datum.new(12), "datum_int", OVERWRITE_PORT)
+    overwrite_helper(ads.add_string_vector, ads.get_port_data_string_vector, datum.VectorString(["baz", "qux"]), "string_vector", OVERWRITE_PORT)
```

Comment:
Just noting that it would be useful for there to be something like `datum.coerce(d)`, which takes `d` and returns the datum of the appropriate type depending on the input Python type (or TypeError if unknown). Also `ads.add` where it is able to coerce the type correctly would be useful. Not sure if these already exist and you are using explicit calls, but if not these would be goods thing to add to Pythonic improvements.

Reply:
Implemented for both ADS and datum in 1c40fdd8f27f207f3620e362328f94bf75ee9e11